text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+}
+
+func (h HealthCheck) Command() []string {
+    return h.command
+}
+
+func (h HealthCheck) Retries() int {
+    return h.retries
+}
+
+func (h HealthCheck) Interval() cdk.Duration {
+    return h.interval
+}
+
+func (h HealthCheck) StartPeriod() cdk.Duration {
+    return h.startPeriod
+}
+
+func (h HealthCheck) Timeout() cdk.Duration {
+    return h.timeout
+}
+
+func renderHealthCheck(hc HealthCheck) CfnTaskDefinition.HealthCheckProperty {
+    return CfnTaskDefintion.HealthCheckProperty{
+        command:     hc.Command(),
+        interval:    hc.Interval(),
+        retries:     hc.Retries(),
+        startPeriod: hc.StartPeriod(),
+        timeout:     hc.Timeout(),
+    }
+}
+```
+
+#### Case 2: Extending Typescript datatype interfaces
+
+For Typescript datatype interfaces that extend another datatype interface, the corresponding interfaces and implementing structs would be generated in
+go, with the extending interface embedding the extended one, e.g.:
+
+```ts
+export interface BaseServiceOptions {
+  readonly cluster: ICluster;
+  readonly desiredCount?: number;
+  readonly serviceName?: string;
+  readonly maxHealthyPercent?: number;
+  readonly minHealthyPercent?: number;
+  readonly healthCheckGracePeriod?: Duration;
+  readonly cloudMapOptions?: CloudMapOptions;
+  readonly propagateTags?: PropagatedTagSource;
+  readonly enableECSManagedTags?: boolean;
+  readonly deploymentController?: DeploymentController;
+}
+
+export interface BaseServiceProps extends BaseServiceOptions {
+  readonly launchType: LaunchType;
+}
+```
+
+As before, the Go translation would have the getter methods (e.g. readonly cluster) defined in a the generated interface, which correspond to
+non-exported struct members in the struct that implements each getter.
+
+```go
+package ecs
+
+type BaseServiceOptionsIface interface {
+    Cluster()                ICluster
+    DesiredCount()           *int
+    ServiceName()            *string
+    MaxHealthyPercent()      *int
+    MinHealthyPercent()      *int
+    HealthCheckGracePeriod() *Duration
+    CloudMapOptions()        CloudMapOptions
+    PropagateTags()          PropagatedTagSource
+    EnableECSManagedTags()   *bool
+    DeploymentController()   DeploymentController
+}
+
+type BaseServicePropsIface interface {
+    serviceOptions           BaseServiceOptionsIface  // embeddeded interface
+    LaunchType()             LaunchType
+}
+
+type BaseServiceOptions struct {
+    cluster                ICluster
+    desiredCount           *int
+    serviceName            *string
+    maxHealthyPercent      *int
+    minHealthyPercent      *int
+    healthCheckGracePeriod *Duration
+    cloudMapOptions        CloudMapOptions
+    propagateTags          PropagatedTagSource
+    enableECSManagedTags   *bool
+    deploymentController   DeploymentController
+}
+
+type BaseServiceProps struct {
+    baseOptions         BaseServiceOptions  // embedded (anonymous?) field
+    launchType          LaunchType
+}
```

Comment:
Should the `BaseServiceOptions` be an anonymous field in this example? Its referred to as an embedded field in the doc below.

Reply:
Did you consider generating these struct values flattened instead of nested? Flattened would make it easier for struct initialization without needing to do nesting. This may help reduce some noise for these options. It looks like the common jsii modeling is to rely heavy on nesting of these  options.

Reply:
Yes, @eladb had also mentioned this [here](https://github.com/aws/aws-cdk-rfcs/pull/206#discussion_r456901209) -- I had some concerns about this necessitating a good deal of boilerplate being generated that might be hard to maintain should the base interface end up being modified.

I was assuming Gophers are used to the idiom of having to be explicit about the embedded field upon initialization (See the linked [example](https://play.golang.org/p/XRpTH3LN6bx))  but I assume the suggestion here would look more like (see: https://play.golang.org/p/rcSOhj0K04H): 

```go
package main

import (
	"fmt"
)

type BaseServiceOptionsIface interface {
	ServiceName() string
}

type BaseServicePropsIface interface {
	BaseServiceOptionsIface // embeddeded interface
	LaunchType() string
}

type BaseServiceOptions struct {
	servicename string
}

type BaseServiceProps struct {
	servicename string //flattened property taken from BaseServiceOptions
	launchtype  string
}

func (p BaseServiceProps) ServiceName() string {
	return p.servicename
}

func (p BaseServiceProps) LaunchType() string {
	return p.launchtype
}

func toString(props BaseServicePropsIface) string {
	return fmt.Sprintf("Service name: %s, Launch type: %s", props.ServiceName(), props.LaunchType())
}

func main() {
	props := BaseServiceProps{"myService", "EC2"}
	fmt.Println(toString(props))
}
```

Reply:
If you go down the wrote of unexported interface methods to ensure customer's aren't broken when JSII datatype properties are added i think you'll be required to do struct embedding of the "base" type.

If you don't use the unexported interface method approach yeah you can use flattened struct generation, but ever struct must generate the flattened member and method as well. Not a major deal, but might have to deal more with conflict resolution if two base types define the same member or methods.

Also if you go down the second approach of flattened generation you cannot prevent customer code failing to compile when CDK adds properties/methods to datatypes.

Reply:
So it seems like the tradeoff here is:
**Flattened**
Pros: more intuitive, neater construction syntax
Cons: requires more code generation, any change to underlying interface would be a breaking change and would require customer to have to reimplement the interface on their own constructs

**Embedded**
Pros: Automatic inheritance of both embedded properties as well as interfaces that the embedded struct implements, meaning adding properties/methods would not be a breaking change for other constructs
Cons: clunkier construction (as shown in the current example)

What you mention with private interface methods [here](https://github.com/aws/aws-cdk-rfcs/pull/206#discussion_r458984612) doesn't totally make sense to me in the context of datatype interfaces, since I would assume all methods on the interface would be public in order to be callable by anyone downstream, though I do like the approach of embedding a struct that our packages would provide that implement a given interface.

Reply:
FYI I've incorporated both approaches (flattening and embedding) in the design doc with examples.